Add detail and clarity around Environment Vars (#9194)
gatsbyjs/gatsby: docs/docs/environment-variables.md

title: "Environment Variables"
---

## Environments and Environment Variables

You can easily provide environment variables to your site.

Note that we need to distinguish in this discussion between variables which are been defined in
special places in order to be used in different deployment environments, and true OS-level 
environment variables that could be used in, for example, command-line calls.

> We'll call the former "Project Environment Variables" and the latter "OS Environment Variables".

Author: Yeah, I got more and more bored of writing the words as I got further through; should have done a pass through and standardised when I was finished.

> We'll call the former "Project Environment Variables" and the latter "OS Environment Variables".

Author: Thanks for cleaning it up for me :)

In both cases we want to be able to access the relavant value of these variable for the environment
we're in.

By default gatsby supports only 2 environments:

 * If you run `gatsby develop`, then you will be in the 'development' environment.
 * If you run `gatsby build` + `gatsby serve`, then you will be in the 'production' environment.
 
If you want to define other Environments then you'll need to do a little more work. See 'Additional Environments' below.

## Accessing Environment Variables in JavaScript

> All of the Project and OS Environment Variables are only directly available at build time, or

Reviewer: ```suggestion
All of the Project and OS Env Vars are only directly available at build time, or
```

when Node.Js is running. They aren't immediately available at run time of the client code; they
need to be actively captured and embedded into our client-side JavaScript.
This is achieved during the build using Webpack's [DefinePlugin](https://webpack.js.org/plugins/define-plugin/).

> Once the Environment Variables have been embedded into the client-side, they are accessible from the

Reviewer: Here I've gone with _environment variables_ when not specifically referring to _OS Env Vars_ or _Project Env Vars_

```suggestion
Once the environment variables have been embedded into the client-side, they are accessible from the
```

global variable `process.env`.

> OS Environment Variables are accessible in Node.js from the same `process.env` global variable.

Reviewer: ```suggestion
OS Env Vars are accessible in Node.js from the same `process.env` global variable.
```

Note that since these variables are embedded at build time, you will need to restart your dev server
or rebuild your site after changing them.

## Defining Environment Variables

#### Client-side JavaScript

> For Project Enviroment Variables that you want to access in client-side browser JavaScript, you can define

Reviewer: ```suggestion
For Project Env Vars that you want to access in client-side browser JavaScript, you can define
```

an environment config file, `.env.development` and/or `.env.production`, in your root folder.
Depending on your active environment, the correct one will be found and its values embedded as EnvVars in the
browser JavaScript.

In addition to these Project Environment Variables defined in `.env.*` files, you could also define
OS Environment Variables. And such OS Env.Vars which are prefixed with `GATSBY_` will become available in
browser JavaScript.

#### Server-side Node.js

Gatsby runs several Node.js scripts at build time, notably `gatsby-config.js` and `gatsby-node.js`.

> OS Env.Vars will already be available when Node is running, so you can add environment variables the

Reviewer: ```suggestion
OS Env Vars will already be available when Node is running, so you can add environment variables the
```

normal ways e.g. by adding environment variables through your hosting/build tool, your OS, or when

> calling gatsby on the command line.

Reviewer: ```suggestion
calling Gatsby on the command line.
```

> In Linux terminals, that is as simple as

Reviewer: Avoiding words like _simple_ and _easy_.

```suggestion
In Linux terminals this can be done with:
```

```
MY_ENV_VAR=foo gatsby develop
```

> In Windows, it's a little harder. https://stackoverflow.com/questions/1420719/powershell-setting-an-environment-variable-for-a-single-command-only documents some options.

Reviewer: ```suggestion
In Windows it's a little more complex. [Check out this Stack Overflow article for some options](https://stackoverflow.com/questions/1420719/powershell-setting-an-environment-variable-for-a-single-command-only documents some options.)
```

> However, the Project Env. Vars that you defined in the `.env.*` files will *NOT* be immediately available

Reviewer: ```suggestion
However, the Project Env Vars that you defined in the `.env.*` files will *NOT* be immediately available
```

in your Node.js scripts. To use those variables, use NPM package [dotenv](https://www.npmjs.com/package/dotenv) to

> examine the active `.env.*` file and attached those values,

Reviewer: ```suggestion
examine the active `.env.*` file and attached values. 
```

> It's already a dependency of gatsby, so simply require it in your `gatsby-config.js` or `gatsby-node.js` like this:

Reviewer: ```suggestion
It's already a dependency of Gatsby, so you can require it in your `gatsby-config.js` or `gatsby-node.js` like this:
```

API_URL=https://example.com/api
```

These variables will be available to your site as `process.env.API_URL`:

```jsx
// usage
render() {
  return (
    <div>
      <img src={`${process.env.API_URL}/logo.png`} alt="Logo" />
    </div>
  )
}
```

> ## Reserved environment variables:

Reviewer: Capitalise heading

```suggestion
## Reserved Environment Variables:
```

> You can not override certain environment variables as some are used internally
> for optimizations during build

- `NODE_ENV`
- `PUBLIC_DIR`

## Additional Environments (Staging, Test, etc)

> As just noted `NODE_ENV` is a reserved environment variable in Gatsby as it is needed by the build system to make key optimizations when compiling React and other modules. For this reason it is necessary to make use of a secondary environment variable for additional environment support, and manually make the Env.Vars available to the client-side code.

Reviewer: Style change? 

```suggestion
As noted above `NODE_ENV` is a reserved environment variable in Gatsby as it is needed by the build system to make key optimizations when compiling React and other modules. For this reason it is necessary to make use of a secondary environment variable for additional environment support, and manually make the environment variables available to the client-side code.
```

> You can define your own OS Environment Variable, to track the active environment, and then to locate the relevant Project Env.Vars to load. The Gatsby itself will not do anything with that OS Env.Var, but you can use it in `gatsby-config.js`.

Reviewer: ```suggestion
You can define your own OS Env Var to track the active environment, and then to locate the relevant Project Env Vars to load. Gatsby itself will not do anything with that OS Env Var, but you can use it in `gatsby-config.js`.
```

> Specifically, you can use `dotenv` and your individual OS Env.Var to locate the `.env.myCustomEnvironment` file, and then use module.exports to store those Project Environment Variables somewhere that the client-side Javascript can access the values (via `graphql` calls.

Reviewer: ```suggestion
Specifically, you can use `dotenv` and your individual OS Env Var to locate the `.env.myCustomEnvironment` file, and then use module.exports to store those Project Env Vars somewhere that the client-side Javascript can access the values (via GraphQL queries).
```

}
```

> This will then load the values from the relevant environment's `.env.*` file and use them in the graphql database and the analytics plugin respectively.

Reviewer: ```suggestion
This will then load the values from the relevant environment's `.env.*` file and make them available via GraphQL queries and the analytics plugin respectively.
```

> Note that `ACTIVE_ENV` could be called anything - it's not used or known about by anything else in Gatsby (as opposed to `NODE_ENV` which is, as previously discussed.

Reviewer: ```suggestion
Note that `ACTIVE_ENV` could be called anything - it's not used or known about by anything else in Gatsby (as opposed to `NODE_ENV` which is, as previously discussed).
```

> Local testing of the `staging` environment is as simple as:

Reviewer: ```suggestion
Local testing of the `staging` environment can be done with:
```
